File: Java/LetterCombination.java
// DP recursive one pass
// use default constructor to build the HashMap
// append current character with former String lists
public class Solution {
    HashMap<Character, String> dict;
    public Solution() {
        dict = new HashMap<Character, String>();
        dict.put('2', "abc");
        dict.put('3', "def");
        dict.put('4', "ghi");
        dict.put('5', "jkl");
        dict.put('6', "mno");
        dict.put('7', "pqrs");
        dict.put('8', "tuv");
        dict.put('9', "wxyz");
    }
    public ArrayList<String> letterCombinations(String digits) {
        ArrayList<String> result = new ArrayList<String>();
        if (digits == null || digits.length() == 0) {
            result.add("");
            return result;
        }
        ArrayList<String> temp = letterCombinations(digits.substring(1, digits.length()));
        char[] currentDigit = dict.get(digits.charAt(0)).toCharArray();
        for (char c:currentDigit) {
            for (String s:temp) {
                result.add(c + s);
            }
        }
        return result;
    }
}


// add - recursive - remove
    public ArrayList<String> letterCombinations(String digits) {
        ArrayList<String> result = new ArrayList<String>();

        if (digits == null) {
            return result;
        }

        Map<Character, char[]> map = new HashMap<Character, char[]>();
        map.put('0', new char[] {});
        map.put('1', new char[] {});
        map.put('2', new char[] { 'a', 'b', 'c' });
        map.put('3', new char[] { 'd', 'e', 'f' });
        map.put('4', new char[] { 'g', 'h', 'i' });
        map.put('5', new char[] { 'j', 'k', 'l' });
        map.put('6', new char[] { 'm', 'n', 'o' });
        map.put('7', new char[] { 'p', 'q', 'r', 's' });
        map.put('8', new char[] { 't', 'u', 'v'});
        map.put('9', new char[] { 'w', 'x', 'y', 'z' });

        StringBuilder sb = new StringBuilder();
        helper(map, digits, sb, result);

        return result;
    }

    private void helper(Map<Character, char[]> map, String digits, 
        StringBuilder sb, ArrayList<String> result) {
        if (sb.length() == digits.length()) {
            result.add(sb.toString());
            return;
        }

        for (char c : map.get(digits.charAt(sb.length()))) {
            sb.append(c);
            helper(map, digits, sb, result);
            sb.deleteCharAt(sb.length() - 1);
        }
    }

File: Python/LetterCombination.py
# DP, first + others, take care of the base case, which is empty list
class Solution(object):
    mapping = {'2':['a','b','c'],
               '3':['d','e','f'],
               '4':['g','h','i'],                   
               '5':['j','k','l'],
               '6':['m','n','o'],
               '7':['p','q','r','s'],
               '8':['t','u','v'],                   
               '9':['w','x','y','z'],                   
               '0':[' ']}    
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        result = []
        if len(digits) == 0:
            return result
        # or deal with len() == 1 specially
        if len(digits) == 1:
            return self.mapping(digits[0])
        others = self.letterCombinations(digits[1:])
        for c in self.mapping[digits[0]]:
            for s in others:
                result.append(c+s)
            if others == []:
                result.append(c)
        return result

# add, recursive, remove
class Solution(object):
    result = []
    one = ''    
    mapping = {'2':['a','b','c'],
               '3':['d','e','f'],
               '4':['g','h','i'],                   
               '5':['j','k','l'],
               '6':['m','n','o'],
               '7':['p','q','r','s'],
               '8':['t','u','v'],                   
               '9':['w','x','y','z'],                   
               '0':[' ']}    
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        self.result = []
        self.one = ''
        if len(digits) < 1:
            return self.result
        self.helper(digits)
        return self.result

    def helper(self, digits):
        if len(digits) == 0:
            self.result.append(self.one[:])
            return
        c = digits[0]
        dicts = self.mapping[c]
        for i in dicts:
            self.one += i
            self.helper(digits[1:])
            self.one = self.one[:-1]
